### PR TITLE
Solver refactoring

### PR DESCRIPTION
**Pull request by mosteo**
This reworks the recursive solver we have into an A*-like sequential one. The advantages are: cleaner implementation, better maintainability, no need to start the search several times to find a good incomplete solution (once we find an incomplete one, it will be decent enough and it means no complete one exists), simpler options for usage.

For users there should be no noticeable changes.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -57,6 +57,9 @@
 [submodule "deps/dirty_booleans"]
 	path = deps/dirty_booleans
 	url = https://github.com/mosteo/dirty_booleans
+[submodule "deps/compare_to_case"]
+	path = deps/compare_to_case
+	url = https://github.com/mosteo/compare_to_case
```
Comment (Fabien-Chouteau) [inline .gitmodules]: I am not too fond of having a dependency project with only one function in it.

File: src/alire/alire-solver.ads [hunk 2]
```diff
@@ -2,71 +2,48 @@ with Alire.Dependencies;
...
-      --  releases. All possible solutions and incomplete subsets are
-      --  eventually explored.
-
-     );
-   --  Allow the solver to further explore incomplete solution space. Each
-   --  value takes more time than the precedent one. All_Incomplete can take
-   --  a veeery long time when many crates/releases must be considered. TODO:
-   --  All these policies can go away once we move from a recursive solver to
-   --  a non-recursive priority-based one.
+   --  When looking for releases within a crate, which one to try first. The
```
Comment (Fabien-Chouteau) [inline src/alire/alire-solver.ads:22]: So what is the strategy for Alire?

**Review by Fabien-Chouteau — APPROVED** [review]
I can't say I fully understand the code but it looks well organized and documented.

I am a bit worried about the dependency on a growing number of  very small libraries.

File: .gitmodules (hunk 1 above)
Comment (mosteo) [reply .gitmodules]: I can see the concern. I was operating from the idea of not having in Alire logic unrelated to its function, but what is the minimum reusable code that makes worthwhile moving it into a new crate is certainly debatable.

I will move that function for now into `Alire.Utils`.

File: src/alire/alire-solver.ads (hunk 2 above)
Comment (mosteo) [reply src/alire/alire-solver.ads:22]: Alire defaults to newest first, unless the global switch `--prefer-oldest` is used, which affects `get`, `search` and `show`. Notably, `alr update` (and automatic updates after manual manifest edition) don't honor it, so I've had for a time the idea of moving it to a setting, so the behavior is globally consistent.

File: src/alire/alire-meta.ads [hunk 3]
```diff
@@ -6,8 +6,8 @@ package Alire.Meta with Preelaborate is
 
    package Working_Tree is
 
-      Commit  : constant String := "unknown";
-      Changes : constant String := "unknown";
+      Commit  : constant String := "c3dade93403aaacec59097e450a22ce83e0ceb0f";
+      Changes : constant String := "dirty";
```
Comment (Seb-MCaw) [inline src/alire/alire-meta.ads:10]: Did you intend to include this?

Comment (mosteo) [reply src/alire/alire-meta.ads:10]: No, thanks for the catch!

**Pull request merged**